doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,266 @@
+- If a module only supports approved services in an approved manner an implicit indicator can be used (e.g. successful completion of a service is an indicator).
+- An approved algorithm is not considered to be an approved implementation if it does not have a CAVP certificate or does not include its required self-tests. (i.e. My interpretation of this is that if the CAVP certificate lists an algorithm with only a subset of key sizes, digests, and/or ciphers compared to the implementation, the differences ARE NOT APPROVED. In many places we have no restrictions on the digest or cipher selected).
+- Documentation is required to demonstrate how to use indicators for each approved cryptographic algorithm.
+- Testing is required to execute all services and verify that the indicator provides an unambiguous indication of whether the service utilizes an approved cryptographic algorithm, security function or process in an approved manner or not.
+- The Security Policy may require updates related to indicators. AWS/google have added a table in their security policy called ‘Non-Approved Algorithms not allowed in the approved mode of operation’. An example is RSA with a keysize of < 2048 bits (which has been enforced by [4]).
+
+Since any new FIPS restrictions added could possibly break existing applications
+the following additional OpenSSL requirements are also needed:
+
+- The FIPS restrictions should be able to be disabled using Configuration file options (This results in unapproved mode and requires an indicator).
+- A mechanism for logging the details of any unapproved mode operations that have been triggered (e.g. DSA Signing)
+- The FIPS restrictions should be able to be enabled/disabled per algorithm context.
+- If the per algorithm context value is not set, then the  Configuration file option is used.
+
+Solution
+--------
+
+In OpenSSL most of the existing code in the FIPS provider is using
+implicit indicators i.e. An error occurs if existing FIPS rules are violated.
+
+The following rules will apply to any code that currently is not FIPS approved,
+but needs to be.
+
+- The fipsinstall application will have a configurable item added for each algorithm that requires a change.
+These options will be passed to the FIPS provider in a manner similar to existing code.
+
+- A user defined callback similar to OSSL_SELF_TEST will be added. This callback
+will be triggered whenever an approved mode test fails.
+It may be set up by the user using
+
+```c
+typedef int (OSSL_CALLBACK)(const OSSL_PARAM params[], void *arg);
+
+void OSSL_INDICATOR_set_callback(OSSL_LIB_CTX *libctx, OSSL_CALLBACK *cb, void *cbarg)
+```
+The callback and/or cbarg can be changed at any time.
+
+- Internally within the FIPS module algorithms the following internal helper
+callback may be called
+
+```c
+int ossl_INDICATOR_callback(OSSL_LIB_CTX *libctx, const char *algtype,
```

Comment:
The libctx argument here is meaningless.  The FIPS provider's internal libctx is the only one available and it's unusable outside.  I'd suggest `PROV_CTX * here` and a wrapper around the callback that converts it: `const OSSL_CORE_HANDLE *handle = CORE_HANDLE(prov_ctx);`

Reply:
Why a callback at all?
It is a very heavy lifting for an application to set up callback (besides the libctx issue).
Other FIPS modules have only one or two modes of operations:
1. Strict: anything not approved just returns an error
2. Relax: Anything not approved exposes an indicator that the application can look at if it cares

And 2. can very simply be implemented with a ctx_get_params() call that queries an indicator parameter.

What good does a full callback system do ?